### PR TITLE
feat: s3 backup store can set status

### LINKED ISSUE #10134: Backup store can set backup status

### PR DESCRIPTION
**Pull request by lenaschoenburg**
The S3 backup store can now set the status of a backup. It does so on-demand when `markAsFailed` is called and during the `save` process where a backup first starts off as `IN_PROGRESS` and is then marked as failed or completed.

`BackupStatus` now contains a `BackupStatusCode`. We will extend the status with more fields later, once we tackle querying the status.

The `save` and `markAsFailed` methods now return a `BackupStatusCode` enum that indicates what status was set last (completed or failed).

 `save` will always return an exceptionally completed future if any of the steps failed. This keeps error handling (responsibility of the backup manager) easy and consistent.

closes #10134

### CODE REVIEW

**Comment by oleschoenburg** [comment]
@deepthidevaki error handling is a bit different to what we last discussed: 

> save will always return an exceptionally completed future if any of the steps failed. This keeps error handling (responsibility of the backup manager) easy and consistent.

I was getting uncomfortable with the approach of just using the status code to indicate failure while writing tests and realizing that the behavior changes depending on where exactly the failure occured - while setting the status to failed or while uploading other objects.

The commit history is a bit long and noisy - let me know if I should squash some of them.

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java [hunk 1]
```diff
@@ -77,8 +88,27 @@ public CompletableFuture<Backup> restore(final BackupIdentifier id) {
   }
 
   @Override
-  public CompletableFuture<Void> markFailed(final BackupIdentifier id) {
-    throw new UnsupportedOperationException();
+  public CompletableFuture<BackupStatusCode> markFailed(final BackupIdentifier id) {
+    return setStatus(id, new Status(BackupStatusCode.FAILED, "Explicitly marked as failed"));
+  }
```
Comment (deepthidevaki) [inline backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java:93]: :thought_balloon: We could also pass the error message as a parameter. BackupManager can then set the reason why it is explicitly marking it as failure.

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java [hunk 2]
```diff
@@ -54,11 +55,21 @@ public S3BackupStore(final S3BackupConfig config, final S3AsyncClient client) {
   }
 
   @Override
-  public CompletableFuture<Void> save(final Backup backup) {
-    final var metadata = saveMetadata(backup);
-    final var snapshot = saveSnapshotFiles(backup);
-    final var segments = saveSegmentFiles(backup);
-    return CompletableFuture.allOf(metadata, snapshot, segments);
+  public CompletableFuture<BackupStatusCode> save(final Backup backup) {
```
Comment (deepthidevaki) [inline backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java:58]: :question: If the future fails with exception, is it useful to return the `BackupStatusCode` here?

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java (hunk 1 above)
Comment (deepthidevaki) [reply backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java:93]: :question: If for any reason the status is already "completed", should we overwrite and mark it as failed?

**Review by deepthidevaki — APPROVED** [review]
:rocket:

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java (hunk 1 above)
Comment (oleschoenburg) [reply backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java:93]: When would the manager actually call this? Depending on the use case, overwriting a "completed" status makes could make sense or not.

File: backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java (hunk 2 above)
Comment (oleschoenburg) [reply backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java]: You are right, I'll omit the status

**Pull request merged**